README.md
<p align="center">
  <img width="40%" height="30%" src="https://github.com/taskcontrols/py-taskcontrol/blob/master/docs/images/logo.jpg">
</p>

# taskcontrol  (py-taskcontrol)
    Create named shared / isolated workflow task controls, and run them with respective before/after middlewares


taskcontrol (py-taskcontrol) is a python library to create tasks in and based on named workflow controls. It allows middlewares before and after each task. taskcontrol can run single or multiple tasks at a task run invocation.

It provides a simple decorator called `workflow` that takes the name, task_instance, task_order, shared, before, after arguments to set up the named workflow controls.

It also provides methods to create a plugin and work with tasks as a module and/or pre-created ordered task list.


# Features

* Create Named task controls (tasks) - instance and isolated
* Allows middlewares before / after each task (data fetch, auth, data save, logging, logout, cleanup, etc)
* Access read-only contexts and results of middlewares/tasks
* Allows Merging two instances of task controls with namespace clash handling
* Run instance, shared, and mix of tasks (individual or all groups)
* In-Development: Allows creating, registering, and using task controls as a plugin


# Installation


##### Command:

* Python

        pip3 install taskcontrol


##### Version:

    In Development Version: 1.2.6 (functional - production ready with plugin and concurrency support, fully tested)
    Current Version: 1.2.5 (functional - production ready with most planned features with MVP, partly tested)
    Previous Version: 1.1.2 (functional - production ready minor issues)
    Previous Version: 1.1.0/1.1.1 (functional - not production ready - minor bug. Please upgrade to v1.1.2)



[Releases](https://github.com/taskcontrols/py-taskcontrol/blob/master/docs/users/releases.md) Lists in short the releases of the package


##### Package Link:
    
    https://github.com/taskcontrols/py-taskcontrol
    https://pypi.org/project/taskcontrol
    
##### Website:
    
    https://taskcontrols.github.io/


# Technical Specifications


##### Requirements:

* Python 3.x and above
* Any OS supporting Python 3.x and above


##### Package Dependencies:

* None


##### Quick Demo:

[demo example - main.py](https://github.com/taskcontrols/py-taskcontrol/blob/master/main.py)


```python



from taskcontrol.workflow import workflow, Tasks

sparrow = Tasks()

def middleware_trees(ctx, result, k, c, d, **kwargs):
    print("Running my Middleware Function: middleware_trees - task items", k, c, d, kwargs)


@workflow(
    name="taskname",
    task_order=1,
    task_instance=sparrow,
    shared=False,
    args=[1, 2],
    kwargs={},
    before=[
        {
            "function": middleware_trees,
            "args": [11, 12],
            "kwargs": {"d": "Before Testing message Middleware "},
            "options": {"error": "next", "error_next_value": ""}
        }
    ],
    after=[
        {
            "function": middleware_trees,
            "args": [13, 14],
            "kwargs": {"d": "After Middleware Testing message"},
            "options": {
                "error": "error_handler",
                "error_next_value": "value",
                "error_handler": lambda err, value: (err, None)
            }
        }
    ],
    log=False
)
def taskone(ctx, result, a, b):
    print("Running my task function: taskone", a, b)


# Run single task
sparrow.run(tasks="taskname")


# Run all tasks
sparrow.run(tasks=["1"])
# sparrow.run(tasks=["taskname", ..., "anothertask"])



```


##### Note:

Though it may support Python version 2.x. However, it has not been tested in 2.x. The Syntax and Features of the library supports Python version 2.x. Use at your own risk.



# Wiki

* [Getting started](https://github.com/taskcontrols/py-taskcontrol/blob/master/docs/users/getting-started.md)
    
    Describes in short the usage of the package

* [taskcontrol `workflow` decorator](https://github.com/taskcontrols/py-taskcontrol/blob/master/docs/users/workflow.md)
    
    Describes how to use the taskcontrol workflow decorator in detail

* [taskcontrol `workflow` decorator argument details](https://github.com/taskcontrols/py-taskcontrol/blob/master/docs/users/workflow-arguments.md)
    
    Describes in detail the arguments of workflow decorator

* [taskcontrol `workflow` before / after argument for middleware declaration](https://github.com/taskcontrols/py-taskcontrol/blob/master/docs/users/workflow-middlewares.md)
    
    Describes creating, defining, and running middlewares

* [taskcontrol `workflow` instance and shared tasks argument](https://github.com/taskcontrols/py-taskcontrol/blob/master/docs/users/workflow-instance-shared-tasks.md)
    
    Describes creating a instance (isolated task) and an shared task (available to all instances)


##### Crazy Hint:
You can also create a simple workflow without taskcontrol using a simple list or nested list and loop through them using a for/while loop and invoke them during looping


```python


# Loop the lists below and invoke the functions 
lst = ["f1", "f2", "f3"]
nest_lst = [["f1", "f2"], "f3", "f4", ["f5"]]


# Use a reducer if you want to send args to next function like below
def test(a,b):
    print(a,b)
    return {"a":a, "b":b}
def tester(a,b):
    print(a,b)
    return None

kwargs_for_first_function_the_its_returns_or_other_value_for_next_func = {"a":"a", "b":"b"}
ls = [kwargs_for_first_function_the_its_returns_or_other_value_for_next_func, test, tester]
import functools 
def red(kwargs_for_first_then_func, p):
    i = p(kwargs.get("a"), kwargs.get("b"))
    return i
functools.reduce(red, ls)


```

# [Todo](https://github.com/taskcontrols/py-taskcontrol/blob/master/.todo)


* e2e and Unit Tests - Add Tests (Structure of package created - to be cleaned after writing tests)
* Allow creating and registering a set of task controls as a plugin
* Add logging system


# Status

* In Active Development (taskcontrol version 1.2.6)


# Support

[Paypal.me/taskcontrols](https://paypal.me/taskcontrols)

[Be a Patreon](https://www.patreon.com/taskcontrols)


# License

The MIT License (MIT) - See [LICENSE](https://github.com/taskcontrols/py-taskcontrol/blob/master/LICENSE) for further details


Copyright © 2020 - till library works

taskcontrols@gmail.com




docs/releases.md
# taskcontrol


# Release 1.2.5

### Features
    Functional, tested release (Major release version)
    
    Change of core functional structure and shared tasks execution in 1.2.5 version
    Read only ctx, and result is compulsary argument which are magically provided
    Ordering of tasks feature functional for all tasks run enabled
    Logger functionality enabled
    Minor bug fix in task run
    Minor doc bugs removed
    Minor support using docs

### BugFix


# Release 1.1.2

    Major bug fix for runtime from version 1.1.1

### Features
### BugFix


# Release 1.1.1

    Minor non-breaking change in arguments (not production ready)


# Release 1.1.0

    Minor non-breaking change in arguments (not production ready)



main.py
# for git development repo
# from taskcontrol.workflow import workflow, Tasks

# for package
# from taskcontrol.workflow import workflow, Tasks
from wf import workflow, Tasks

# Instance of tasks and apis object
# Every instance will store it own list of tasks
#       with their before/after middlewares
sparrow = Tasks()


def nesttree(ctx, result, k, c, d, **kwargs):
    print("Running my Middleware Function: nesttree - task items", k, c, d, kwargs)


@workflow(
    name="taskname",
    task_order=1,
    task_instance=sparrow,
    shared=False,
    args=[1, 2],
    kwargs={},
    before=[
        # before middleware order followed will be of the list sequence
        {
            "function": nesttree,
            "args": [11, 12],
            "kwargs": {"d": "Before Testing message Middleware "},

            # options { error : str,  error_next_value: Object, error_handler: function }
            #
            # error { str }: [next, error_handler, exit]
            # error_handler { function }
            # error_next_value { object }
            #
            # Usage:
            # "options": {"error": "next", "error_next_value": "value"}
            # "options": {"error": "exit"}
            # "options": {
            #    "error": "error_handler", error_handler: func, "error_next_value": "value"
            #    }

            "options": {"error": "next", "error_next_value": ""}
        }
    ],
    after=[
        # after middleware order followed will be of the list sequence
        {
            "function": nesttree,
            "args": [13, 14],
            "kwargs": {"d": "After Middleware Testing message"},
            "options": {
                "error": "error_handler",
                "error_next_value": "value",

                #
                # Default error_handler implementation used internally, if no
                #           error_handler is provided
                #
                # Implementation One:
                #   if error_next_value defined
                #       lambda err, value: (err, error_next_value)
                # Implementation Two:
                #   if error_next_value not defined
                #       lambda err, value: (err, None)
                #
                # Returning the two value tuple in error_handler implementation is compulsary
                #       err is the error that occurred
                #       error_next_value is error_next_value provided in options
                #

                "error_handler": lambda err, value: (err, None)
            }
        }
    ],
    log=False
)
def taskone(ctx, result, a, b):
    print("Running my task function: taskone", a, b)


# Example two for decorator usage
@workflow(name="tasktwo",
          task_instance=sparrow,
          task_order=2,
          shared=False,
          args=[1, 2],
          kwargs={},
          # Declare before/after as an list or an object (if single middleware function)
          before={
              "function": nesttree,
              "args": [21, 22],
              "kwargs": {"d": "Before Testing message"},
              "options": {"error": "next", "error_next_value": ""}
          },
          after=[],
          log=False
          )
def tasktwo(ctx, result, a, b):
    print("Running my task function: tasktwo", a, b)
    return a, b


# INVOKE BELOW WHERE NEEDED
# Example: Within some other function


# Run all tasks
# Multiple Workflow Tasks run
sparrow.run(tasks=["1"])
sparrow.run(tasks="1")


# Run all shared tasks
# Shared Workflow Tasks run
sparrow.run(tasks=["shared:1"])
sparrow.run(tasks="shared:1")


# Multiple Workflow Tasks run
run_1 = sparrow.run(tasks=["shared:taskname", "tasktwo"])
# print("sparrow.ctx ", sparrow.ctx)
print("run_1", run_1)


# Run Tasks run with mix of shared
# Multiple Workflow Tasks run with mix of shared
sparrow.run(tasks=["taskname", "tasktwo", "shared:taskname"])


# Single Workflow Tasks run
run_2 = sparrow.run(tasks="shared:taskname")
print("run_2", run_2)


# Run Tasks run with shared task
# Single Workflow Tasks run for shared task
sparrow.run(tasks="shared:taskname")

